docs/enhancements/20190815-zone-aware-routing.md

```diff
@@ -45,7 +48,54 @@ Arguably inter-zone network latency should also be better than cross zone.
+* This feature inherently assumes that endpoints are distributed across zones in a way that they can handle all the traffic from ingress-nginx pod(s) in that zone
+* This feature will be relying on https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#failure-domainbetakubernetesiozone, it is not this KEP's goal to support other cases
+
+## Proposal
+
+The idea here is to have controller part of ingress-nginx to (1) detect what zone its current pod is running in and (2) detect the zone for every endpoints it knows about.
+After that it will post that data as part of endpoints to Lua land. Then Lua balancer when picking an endpoint will try to pick zone-local endpoint first and
+if there is no zone-local endpoint then it will fallback to current behaviour.
+
+This feature at least in the beginning should be optional since it is going to make it harder to reason about the load balancing and not everyone might want that.
+
+**How does controller know what zone it runs in?**
+We can have the pod spec do pass node name using downward API as an environment variable.
+Then on start controller can get node details from the API based on node name. Once the node details is obtained
+we can extract the zone from `failure-domain.beta.kubernetes.io/zone` annotation. Then we can pass that value to Lua land through Nginx configuration
+when loading `lua_ingress.lua` module in `init_by_lua` phase.
+
+**How do we extract zones for endpoints?**
+We can have the controller watch create and update events on nodes in the entire cluster and based on that keep the map of nodes to zones in the memory.
+And when we generate endpoints list, we can access node name using `.subsets.addresses[i].nodeName`
+and based on that fetch zone from the map in memory and store it as a field on the endpoint.
+__This solution assumes `failure-domain.beta.kubernetes.io/zone`__ annotation does not change until the end of node's life. Otherwise we have to
+watch update events as well on the nodes and that'll add even more overhead.
+
+Alternatively, we can get the list of nodes only when there's no node in the memory for given node name. This is probably a better solution
+because then we would avoid watching for API changes on node resources. We can eagrly fetch all the nodes and build node name to zone mapping on start.
+And from thereon sync it during endpoints building in the main event loop iff there's no entry exist for the node of an endpoint.
+This means an extra API call in case cluster has expanded.
+
+**How do we make sure we do our best to choose zone-local endpoint?**
+This will be done on Lua side. For every backend we will initialize two balancer instances: (1) with all endpoints
+(2) with all endpoints corresponding to current zone for the backend. Then given the request once we choose what backend
+needs to serve the request, we will first try to use zonal balancer for that backend. If zonal balancer does not exist (i.e there's no zonal endpoint)
```

Comment:
**Below are my raw thoughts, will edit and simplify once I digest it myself**


This logic inherently assumes that endpoints are distributed across zones in proportion to ingress-nginx pods distribution and requests are evenly distributed across ingress-nginx pods. While the second part of the assumption is mostly the case in practice it might be very difficult to ensure the first part.

What we can do instead we can have ingress-nginx to dynamically distribute traffic based on endpoints distribution across zones. Given `N` requests, distributed evenly among `M` ingress-nginx pods then each pod gets `N / M` requests. Let's say there are 3 zones. Since we can not guarantee ingress-nginx pods are scheduled evenly across zones, let's say zone `i` has `z_i*M/3` ingress-nginx pods. That means it will get `(N / M) * (z_i * M / 3)` requests. While under perfect distribution it would be `(N / M) * (M / 3)`. Reducing the formulas we get `N*z_i/3` and `N/3`. That means zone `i` gets `N * (z_i - 1) / 3` more requests. Now if all the endpoints would be evenly distributed across zones we would do strict zone-local routing if `N * (z_i - 1) / 3 < 0`. Otherwise we would route to other zones with the probability of `(N * (z_i - 1) / 3) / (N / 3)`. That way we guarantee that requests are load balanced evenly across endpoints across zones. But if like ingress-nginx pods, we can not ensure endpoints to be evenly distributed across zones then we have to apply a similar calculation. Now assume there are `E` endpoints. Evenly distribution would mean each zone has `E / 3` endpoints. But let's assume zone `i` has `e_i * E / 3` endpoints. Now similarly we can say zone `i` will have `e_i * E / 3 - E / 3` more endpoints. Which is `E * (e_i - 1) / 3`. Now if 

`N * (z_i - 1) / 3 <= 0` and `E * (e_i - 1) / 3 >= 0` that means there are enough endpoints to handle zone-local traffic

`N * (z_i - 1) / 3 <= 0` and `E * (e_i - 1) / 3 < 0` then if `z_i - e_i < 0` with probability of `z_i / e_i` proxy to other zones otherwise keep requests local otherwise

`N * (z_i - 1) / 3 > 0` and `E * (e_i - 1) / 3 <= 0` that means now enough zone local endpoints, we need to rout to other zones

`N * (z_i - 1) / 3 > 0` and `E * (e_i - 1) / 3 > 0` then if `z_i - e_i < 0` then keep requests local otherwise with probability of `z_i / e_i` proxy to other zones

Reply:
> Since we can not guarantee ingress-nginx pods are scheduled evenly across zones

Yes, you can do that using nodeAffinity to distribute the ingress controller pods between zones https://kubernetes.io/docs/concepts/configuration/assign-pod-node/

> Now if all the endpoints would be evenly distributed across zones we would do strict zone-local routing

Same thing for the apps. Without nodeAffinity there is no way to ensure the distribution.

Reply:
But does node affinity ensure distribution though? Does it not just do best effort?

Reply:
You can force it with `requiredDuringSchedulingIgnoredDuringExecution` https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#node-affinity

Reply:
I think we can come up with a heuristic as I was trying above in my comment that let's us to dynamically detect the proportion of pods distribution across zones and based on that decide whether request should be routed locally or to an endpoint in the other zone. It boils down to following logic

- if ingress-nginx pod that received request is in the zone that have less ingress-nginx pods in total
  - if there's more service endpoints in this zone then always do zone-local proxying because you know there are more than enough endpoints in this zone
  - otherwise if there are less endpoints in this zone then we have to calculate how much less endpoints there are in this zone in proportion to how much less traffic we get in this zone. If we still get more traffic in proportion then we route out excess traffic to other zone otherwise do zone-local proxying

- if ingress-nginx pod that received request is in the zone that have more ingress-nginx pods in total
  - if there's more service endpoints in this zone then we have to calculate how much more endpoints there are in this zone in proportion to how much more traffic we get in this zone. If we get even more traffic then we route out excess traffic to other zone otherwise do zone-local proxying
  - if there's less service endpoints in this zone then we again calculate the proportion and based on that with some probability we route out to other zones otherwise keep it zone-local

Reply:
> You can force it with requiredDuringSchedulingIgnoredDuringExecution https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#node-affinity

I wonder how much overhead this adds on scheduler when you have i.e over 500 replicas

Reply:
Envoy also has zone aware routing feature. However what I'm proposing in this comment is significantly different than the way Envoy implements it. Envoy spills over traffic to other zones only when there are capacity issue of some sort, in other words reactively.
But the above heuristic ensures that we split traffic evenly across endpoints even if ingress-nginx and app pods are not evenly distribution across zones.

I'd like to also highlight that this is not addressing zonal failures.

Reply:
Actually Envoy does something similar too according to https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/zone_aware